students/Gabe_Lamberth/lesson05/mailroom3.py

```diff
@@ -0,0 +1,153 @@
+#!/usr/bin/env python3
+
+from pathlib import Path
+from random import randint
+
+# Global data structure
+patrons = {'Emma Watson': [50.00, 150.00, 2000.00, 3000.00],
+           'Carolyn Wilson': [5000.25, 2500.00],
+           'Andres Dominguez': [5015.00, 200.50, 35.50]}
+
+
+# Called from menu_dictionary in main()
+def generate_report():
+    header = ('Donor Name', 'Total Given', 'Num Gifts', 'Average Gift')
+    print('{:25} |{:^15} |{:^15} |{:^15}'.format(*header))
+    print('-' * 75)
+    """
+    Combined print statement with dictionary comprehension expression
+
+        for key, value in patrons.items():
+                print(f"{key:<30} ${sum(value):^,.2f}  {len(value):^20}  ${sum(value)/len(value):,.2f}")
+    """
+    print(''.join({f"{k:<30} ${sum(v):^,.2f}  {len(v):^20}  ${sum(v)/len(v):,.2f}\n" for (k, v) in patrons.items()}))
+    print('-- Report Finished --\n')
+
+
+# Called from send_thanks function
+def send_email(choice, amount):
+    print(f"Thank you {choice} for your generous ${amount:,.2f} donation")
+    print('-- Email Sent --\n')
+
+
+# Called when user wants to see the list of donors
+def display_donors():
+    print('Showing list of patrons in the database:')
+    """
+    Combined print statement with dictionary comprehension expression 
+
+        for key, value in patrons.items():
+                print(key, end=', ')
+    """
+    print('\n'.join({k for k in patrons.keys()}))
+
+
+# Used to help add_donor menu validation for dictionary function call
+def silent_quit():
+    pass
+
+
+# Called from menu_dictionary
+def add_donor():
+    # donor menu used to help validate choices
+    donor_menu = {'1': display_donors, 'quit': silent_quit}
+
+    while True:
+        choice = input("""
+        Type 1 - To display a list of donor's.
+        Type quit - to exit to main menu.
+        Enter value here >>  """)
+        try:
+            donor_menu[choice]()
+        except KeyError:
+            print("Not a valid choice, please try again")
+        else:
+            if choice == 'quit':
+                break
+            else:
+                first, last = input("\nPlease enter the new Donor's First and Last name:  ").split()
+                name = f'{first} {last}'.title()
+                if name in patrons.keys():
+                    print("Donor's name already exists")
+                    add_donor()
```

Comment:
careful here, when a function calls itself you can get into some very odd behavior.  Best to just use a `continue`